### ISSUE TITLE
Add new scoped callbacks

### ISSUE THREAD

**Issue by reevn** [kotest/kotest#1584]
`beforeTest`/`afterTest` is being executed additional times for containers (i.e. `describe` block) before then being executed again for the tests in the container.

Example test class
```kotlin
class SomeSpec: DescribeSpec({
  beforeTest {
    println("beforeTest 1")
  }

  afterTest {
    println("afterTest 1")
  }

  describe("some scope") {
    beforeTest {
      println("beforeTest 2")
    }

    it("test A") {
      println("test A")
    }

    it("test B") {
      println("test B")
    }

    describe("some deeper scope") {
      it("test C") {
        println("test C")
      }

      it("test D") {
        println("test D")
      }
    }
  }
})
```

Output (4.2.0.393-SNAPSHOT, as another listener fix is included there):

```
beforeTest 1

// test case "A" start
beforeTest 1
beforeTest 2
test A
afterTest 1
// test case "A" end

// test case "B" start
beforeTest 1
beforeTest 2
test B
afterTest 1
// test case "B" end

beforeTest 1
beforeTest 2

// test case "C" start
beforeTest 1
beforeTest 2
test C
afterTest 1
// test case "C" end

// test case "D" start
beforeTest 1
beforeTest 2
test D
afterTest 1
// test case "D" end

afterTest 1
afterTest 1
```

**Basically everything not in-between the start/end comments is not just unnecessary, it messes up certain test cases where the typical "mock in beforeTest and cleanup in afterTest" pattern is used.**

___

### Scenario

- mocks are created at the top level scope outside of `beforeTest 1`
- `beforeTest 1` defines behaviour for mocks/listeners and creates a fresh instance of the class under test
- `afterTest 1` is supposed to clear all the mocks/reset the listeners

As you can see in the beginning of the output, `beforeTest 1` is executed twice in a row without `afterTest 1` being executed in between. This leads to problems as `beforeTest 1` now created two instances of the class under test without mocks/listeners being reset in-between.

I think `afterTest 1` not being executed here is a bug, but I also don't really understand why listeners like before**_Test_** are executed additionally for containers before executing them again for the actual tests.

I already identified which lines in the project would need to be changed for every execution outside the `start/end` comments to not occur, but that fix seems more like fighting the symptom and not the root cause.

Maybe the real problem is that containers are considered tests (and therefore add to the test count etc.), which should also not be the case?

**Comment by sksamuel** [comment]
Ok, so in Kotest each scope is considered a Test. So if you have a describe block with 2 nested it's, you have 3 tests.

beforeTest is a callback that is invoked before each test, so in this case 3 times (the behavior you're seeing).

The reason for this is somewhat historical. JUnit platform has the concept of containers and tests, but gradle never did (still doesn't I don't think) support this properly. In fact in gradle 4.x, it would just outright hang if you tried to use containers in this way. It's gotten better since but it's still rooted in the land of classes for containers and methods for tests.

The beforeTest behavior isn't going to change now, as that would break things for far too many people. However we could introduce extra callbacks to give finer grained control.

Perhaps `beforeRootTest` which only fires once per top level test.
And `beforeLeafTest` which is only fired for outer test cases.

I don't particularly like the nomenclature there, but that functionality would be easy to add.

**Comment by reevn** [comment]
Thanks for the explanation.

What do you think about adding a flag for the desired behaviour I described that one would use in the Project/Spec config?

Something like

`executeListenersForContainers`, which will have `true` as a default where it would behave as of now. If one sets it to `false`, relevant listeners will only be called for `TestType.Test`.

Implementing this seems very easy from looking into the code.

Basically `beforeTest`/`afterTest` just needs an additional check before executing the listener:

`if (!config.executeListnersForContainers && testCase.type == TestType.Container) return`.

I would be happy to open a PR for this so our team can finally make the switch to Kotest 🙂.

**Comment by sksamuel** [comment]
I would rather add another callback than a flag, as this doesn't seem like
behavior you'd ever want to "flip" about, which is why you usually have
flags.
In other words, once the test is written a certain way it must always run
that way to be correct.

That said, if you feel the flag is better I'm open to persuasion.

**Comment by reevn** [comment]
I personally think the flag is better as I think that most people who look into this would agree that this is how it should be in a perfect world, but as we can't break existing behaviour a flag would be the first step towards "correction".

Every parent `beforeTest` needs to be run before a test itself is executed. This is the typical expected behaviour of all testing frameworks I've seen so far. All `beforeTest`s are already executed like that, but additionally they are also executed when a new container is "started up" for the first time. This is the behaviour that I would not consider correct or expected in the classical sense, but I understand your point about how it has to be this way and that it can not be changed immediately for fear of breaking changes.

A flag would not break existing tests, but people who know about this can switch it on to have the typical `BeforeEach` functionality like one might be used to from other frameworks. I personally think this should become standard in some major version release, as I honestly don't think most people are even aware that these listeners are called twice before every first test of a container. They probably just see no test failing and don't investigate this further.

I can't think of a use-case where you would want your `beforeTest`s to be randomly executed twice before the first test of a container. The second execution would do the code the first execution did anyway immediately after.

Adding new callbacks to be used for the correct usage of `beforeEach` would strongly dilute the meaning of what is expected of `beforeTest` in my opinion and overcomplicate matters further.

In summary:
- I think the current default behaviour is not what the majority of people would expect of the function, as it differentiates itself from all other testing frameworks.
- I think the flag would be a first step in the right direction to correct this, where people can opt-in right now. In the future I would say this should become the default
- I don't see the use-case of executing some `beforeTest` twice when the first test of a container is executed. The second execution would execute the same code anyway, why the need to do it twice?

In a perfect world I would imagine these listeners as the bread and butter of how to write clean nested tests:

- `beforeProject` & `afterProject` for code that needs to run before ALL specs and after ALL specs
- `beforeSpec` & `afterSpec` for code that needs to run before a specific spec is started and after a specific spec is finished
- `beforeScope` & `afterScope` for code that needs to run before all containers/tests inside a container and after all containers/tests inside a container (**This one doesn't exist afaik**)
- `beforeTest` & `afterTest` for code that needs to run before _each_ test in a container and after _each_ test in a container (and child containers)

**Comment by reevn** [comment]
For completeness, here is the output with the `TestType` that triggered the listener included:

```
beforeTest 1 - Container
beforeTest 1 - Test
beforeTest 2 - Test
test A
afterTest 1 - Test
beforeTest 1 - Test
beforeTest 2 - Test
test B
afterTest 1 - Test
beforeTest 1 - Container
beforeTest 2 - Container
beforeTest 1 - Test
beforeTest 2 - Test
test C
afterTest 1 - Test
beforeTest 1 - Test
beforeTest 2 - Test
test D
afterTest 1 - Test
afterTest 1 - Container
afterTest 1 - Container
```

This shows that every `Container` execution is meaningless in the best case and even error-prone in the worst case. This behaviour seems like a lose-lose situation with nothing to gain from anybody, but I'm happy to ponder over other opinions 🙂.

**Comment by sksamuel** [comment]
> I can't think of a use-case where you would want your `beforeTest`s to be randomly executed twice before the first test of a container. The second execution would do the code the first execution did anyway immediately after.

I wouldn't say the beforeTest's are randomly executed. The behavior is defined and consistent - beforeTest is executed before every scope. This is just not what you want in your case.

There are people who want this behavior. They nest things in all kinds of weird and wonderful ways. I've seen users ask questions and in the code snippets they post, they are resetting / testing things at every level.

> In summary:
> 
> * I think the current default behaviour is not what the majority of people would expect of the function, as it differentiates itself from all other testing frameworks.

I agree with you, most users probably think of the leaf scope as the "test".

> * I think the flag would be a first step in the right direction to correct this, where people can opt-in right now. In the future I would say this should become the default

We can't change the default behavior as too many people use this library. We'd have to do a full deprecation cycle, remove, and replace, and that would take years.

> Adding new callbacks to be used for the correct usage of `beforeEach` would strongly dilute the meaning of what is expected of `beforeTest` in my opinion and overcomplicate matters further.

We have these discussions a lot (along with whether single instance is the right default), everyone has their own opinion of how it should work, and no one is wrong. Part of this is because Kotest is very permissive in it's layout flexibility - some would say too permissive.

We'll get you the functionality you want, I'm committed to supporting everyone's use cases, but backwards compatibility is a must.

> 
> In a perfect world I would imagine these listeners as the bread and butter of how to write clean nested tests:
> 
> * `beforeProject` & `afterProject` for code that needs to run before ALL specs and after ALL specs
> * `beforeSpec` & `afterSpec` for code that needs to run before a specific spec is started and after a specific spec is finished
> * `beforeScope` & `afterScope` for code that needs to run before all containers/tests inside a container and after all containers/tests inside a container (**This one doesn't exist afaik**)
> * `beforeTest` & `afterTest` for code that needs to run before _each_ test in a container and after _each_ test in a container (and child containers)

I think your naming here is better than what we have, and if we were starting again, this would be perfect. We can add beforeScope and afterScope as parallels to beforeTest and afterTest. We could add beforeContainer and afterContainer to only run at branch scopes. We could add beforeTestCase and afterTestCase to do only the outer tests.

> I personally think the flag is better as I think that most people who look into this would agree that this is how it should be in a perfect world, but as we can't break existing behaviour a flag would be the first step towards "correction".

The reason I am against a flag is because it will make you happy, but no one else will find it. It'll end up being hidden away in some documentation page. Whereas with callbacks you will see them in the IDE immediately. New users will look at the options and work out which is right for them.

**Comment by reevn** [comment]
Thanks for the detailed answer.

> The reason I am against a flag is because it will make you happy, but no one else will find it. It'll end up being hidden away in some documentation page. Whereas with callbacks you will see them in the IDE immediately. New users will look at the options and work out which is right for them.

Very reasonable, I can see the general dislike for flags.

> I wouldn't say the beforeTest's are randomly executed. The behavior is defined and consistent - beforeTest is executed before every scope. This is just not what you want in your case.

> There are people who want this behavior. They nest things in all kinds of weird and wonderful ways. I've seen users ask questions and in the code snippets they post, they are resetting / testing things at every level.

I would fully agree with this and find it wonderful that you guys are so open to enable everyone to write tests in the way they want. That is awesome!

In this case I'm not sure that we see the same situation though. The `beforeTest` of the user you would describe would still be executed, it's not gone. It's just not executed an additional time right before it is executed again.

So the transformation with this change or flag would be:

from
```
beforeTest 1
beforeTest 1
beforeTest 2
testCase
```

to
```
beforeTest 1
beforeTest 2
testCase
```

The code of any user of this library would still be executed, it just wouldn't be executed twice immediately in a row. No matter how deep down the first real testCase is, all the `beforeTests` of all scopes that come before will always be executed before the testCase is executed, we're just getting rid of the meaningless duplicate ones.

Aren't there only two scenarios for a `beforeTest` block a user could write?:
- The user-written `beforeTest` is idempotent, so no matter how often it is called in succession, it will always result in the same thing. In this case nothing changes by removing the duplicate callback executions, as all `beforeTest` will still be called exactly once before each testCase
- The user-written `beforeTest` is not idempotent, so multiple calls result in ambiguous or even faulty scenarios. In this case the current behaviour is messing up tests, as the same callback is sometimes executed twice before a test (every first test of a scope), and sometimes just once (every other test after the first one of a scope), so some tests have a different initial state than others.

In both scenarios everybody benefits from not executing `beforeTest` multiple times before a given testCase. Am I missing a setup where the described scenario above is not true? Thanks for taking the time to discuss this!

**Comment by sksamuel** [comment]
> In both scenarios everybody benefits from not executing `beforeTest` multiple times before a given testCase. Am I missing a setup where the described scenario above is not true? Thanks for taking the time to discuss this!

I've seen people do things like this:

```
class FooTest : FunSpec() {
   init {
      beforeTest {
         topic = ... // create new kafka topic
      }

      context("something should do x") {
         // test here
         // confirm kafka topic has message a
         context("and then do Y") {
            // test here
            // confirm kafka topic is empty
            test("and then do Z") {
               // last part
               // confirm kafka topic has message b
            }
         }
      }
   }
}
```

They're using each nested scope as like a "follow on" test, which actually has isolated parameters. I think your change would break this. Not the way I'd do it, and it could certainly be changed to not require the beforeTest, but we can't break the current status quo :)

But I'm in agreement with you on the (lack of) intuitive naming in the current setup. If you can come up with a good set of new callback names then we can:

a) add the new callbacks with more ergonomic nomenclature
b) (optionally) deprecate the beforeTest names.

Some suggestions listed in this thread so far:

`beforeTestCase`
`beforeScope`
`beforeContainer`
`beforeEach`

I think your only objection to this solution is that you like the name `beforeTest` and would prefer that?

**Comment by reevn** [comment]
Thanks for providing the example. I can see how these kinds of tests would not be possible if the behaviour of `beforeTest` right now would be changed, although I think there are test styles that would be better fitting for these kinds of "follow on" tests instead of "misusing" contexts as test cases, but it is how it is 🙂.

> I think your only objection to this solution is that you like the name beforeTest and would prefer that?

A working solution where we can just change all occurrences of `beforeTest` in our codebase with some new callback that has a somewhat meaningful name would be totally fine. For the "good of the project" I think it would be good to give the callbacks another pass when it comes to naming/behaviour, no matter how many years down the line the next major version is 🙂.

I think for the use case where a callback is needed that is only executed before/after `TestType.Test`, something like `beforeEach`/`afterEach` would be nice, as that is how that exact functionality is called in most other frameworks. Might be a bit confusing for people what the difference between `beforeTest` and `beforeEach` would be though.

**Comment by sksamuel** [comment]
> Thanks for providing the example. I can see how these kinds of tests would not be possible if the behaviour of `beforeTest` right now would be changed, although I think there are test styles that would be better fitting for these kinds of "follow on" tests instead of "misusing" contexts as test cases, but it is how it is .

I guess it's personal preference as to whether it's misusing it or not I suppose.

> I think for the use case where a callback is needed that is only executed before `TestType.Test`, something like `beforeEach` would be nice, as that is how that exact functionality is called in most other frameworks. Might be a bit confusing for people what the difference between `beforeTest` and `beforeEach` would be though.

If beforeTest was deprecated it would be clear.

`beforeEach` - A callback executed before each "leaf" TestCase. In Kotest a leaf test is the outmost test scope and has TestType.Test

`beforeContainer` - A callback executed before each "container" TestCase. In Kotest a container test are the inner test copes (that contain other scopes) and have TestType.Container

`beforeAny` | `beforeScope` | `beforeTest` - A callback executed for both TestType.Test and TestType.Container tests.

**Comment by sksamuel** [comment]
Alternatively, we could support:

````kotlin
beforeTest(TestType.Test) {
  println("Runs before tests only")
}
```

````kotlin
beforeTest(TestType.Container) {
  println("Runs before containers only")
}
```

**Comment by reevn** [comment]
> beforeEach - A callback executed before each "leaf" TestCase. In Kotest a leaf test is the outmost test scope and has TestType.Test

> beforeContainer - A callback executed before each "container" TestCase. In Kotest a container test are the inner test copes (that contain other scopes) and have TestType.Container

👍 

> beforeAny | beforeScope | beforeTest - A callback executed for both TestType.Test and TestType.Container tests.

For me `scope` and `containers` were synonyms, so I wouldn't expect `beforeScope` to be executed every time before `TestType.Test`. `beforeAny` is actually pretty nice for this and complements the other two new callbacks well.

> Alternatively, we could support:

This would require boilerplate every time someone writes a `beforeTest`. Additionally this is already possible via adding one line:

```kotlin
beforeTest {
  if (it.type == TestType.Container) return@beforeTest
  // code that should only be executed before leaf tests
}
```

**Comment by sksamuel** [comment]
If you want to make a PR feel free, otherwise I will get around to adding them this weekend.

**Comment by reevn** [comment]
So to summarize (if I have the time for the PR before the weekend):

- new `beforeEach` for `TestType.Test`
- new `beforeContainer` for `TestType.Container`
- new `beforeAny` for `TestType.Test` and `TestType.Container` (same as `beforeTest` right now)
- deprecate `beforeTest` and nudge towards the other functions or `beforeAny` in the deprecation description

If that's all I think I can get to it before the weekend.

**Comment by sksamuel** [comment]
Yes, but without the deprecate part while I garner feedback from other users.
